### PR TITLE
[M2P395] Billing address is invalid if it was created or changed on native checkout page

### PR DESCRIPTION
**Pull request by mikeshutov-bolt**
# Description
Subscribing to billing address change and creating a new order token for the updated billing address. Also using cartBarrier instead of BoltState.boltcart for configure call to make sure we have the latest cart values.

Fixes: [M2P395](https://boltpay.atlassian.net/browse/M2P-395) - Billing address is invalid if it was created or changed on native checkout page

#changelog (WIP)[M2P395] Billing address is invalid if it was created or changed on native checkout page

# Type of change

- [x] Bug fix (change which fixes an issue)
- [ ] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
Please validate that you have tested your change in at least one of the following areas:

- [x] Successfully tested locally (or docker image)
- [ ] Successfully tested on a staging or sandbox server
- [ ] Successfully tested on a merchant's staging server

# For PR Reviewer 
- [ ] Reviewed unit tests to make sure we are using real components rather than mocks as much as possible?
- [ ] For any major change (observer, new Bolt feature, core Magento interaction) we must add a feature switch, did you verify this?

# Checklist:

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have created or modified unit tests to sufficiently cover my changes.
- [ ] I have added my Jira ticket link and provided a changelog message.

### CODE REVIEW

**Review by njfaries — APPROVED** [review]

File: view/frontend/templates/js/replacejs.phtml [hunk 1]
```diff
@@ -1485,6 +1489,15 @@ if (!$block->isOnPageFromWhiteList() && !$block->isMinicartEnabled()) { return;
             'jquery',
             'Magento_Checkout/js/model/quote',
             ], function ($, quote) {
+                var prevAddress;
+                //If billing address has been changed we need to recreate the order token.
+                quote.billingAddress.subscribe(function(newAddress){
+                    if((!newAddress ^ !prevAddress || prevAddress.getKey() !== newAddress.getKey())){
```
Comment (vitaliyreznikov) [inline view/frontend/templates/js/replacejs.phtml:1495]: can we save only prevAddress.key?

**Review by vitaliyreznikov — APPROVED** [review]
Looks good, one non-blocking proposal

File: view/frontend/templates/js/replacejs.phtml [hunk 2]
```diff
@@ -1007,24 +1007,28 @@ if (!$block->isOnPageFromWhiteList() && !$block->isMinicartEnabled()) { return;
                             if (BoltState.cartRestricted) {
                                 if (popUpOpen) reject(new Error(data.message));
                                 hintBarrier.resolve(hints);
+                                cartBarrier.resolve(cart);
```
Comment (danac-gs) [inline view/frontend/templates/js/replacejs.phtml:1010]: Where is this `cart` variable defined? I couldn't find it.

File: view/frontend/templates/js/replacejs.phtml [hunk 3]
```diff
@@ -1007,24 +1007,28 @@ if (!$block->isOnPageFromWhiteList() && !$block->isMinicartEnabled()) { return;
...
                                 return;
                             }
 
                             if (data.status !== 'success') {
                                 reject(new Error(data.message || 'Network request failed'));
                                 hintBarrier.resolve(hints);
+                                cartBarrier.resolve(cart);
```
Comment (danac-gs) [inline view/frontend/templates/js/replacejs.phtml:1017]: Same here. Where is this `cart` variable defined?

File: view/frontend/templates/js/replacejs.phtml [hunk 4]
```diff
@@ -1007,24 +1007,28 @@ if (!$block->isOnPageFromWhiteList() && !$block->isMinicartEnabled()) { return;
...
                                 return;
                             }
 
                             if (!data.cart) {
                                 reject(new Error('The cart info is missing.'));
                                 hintBarrier.resolve(hints);
+                                cartBarrier.resolve(cart);
```
Comment (danac-gs) [inline view/frontend/templates/js/replacejs.phtml:1024]: The same. Where is this `cart` variable defined?

File: view/frontend/templates/js/replacejs.phtml [hunk 5]
```diff
@@ -1007,24 +1007,28 @@ if (!$block->isOnPageFromWhiteList() && !$block->isMinicartEnabled()) { return;
...
                                 return;
                             }
 
                             if (!data.cart.orderToken) {
                                 reject(new Error('The cart is empty.'));
                                 hintBarrier.resolve(hints);
+                                cartBarrier.resolve(cart);
```
Comment (danac-gs) [inline view/frontend/templates/js/replacejs.phtml:1031]: And here. Where is this `cart` variable defined?

**Review by danac-gs — CHANGES_REQUESTED** [review]
Please check the comments.

File: view/frontend/templates/js/replacejs.phtml (hunk 1 above)
Comment (mikeshutov-bolt) [reply view/frontend/templates/js/replacejs.phtml]: I found a different solution for this while checking whole code fixed the old billingAddress listener

File: view/frontend/templates/js/replacejs.phtml (hunk 2 above)
Comment (mikeshutov-bolt) [reply view/frontend/templates/js/replacejs.phtml]: I ended up redoing this and finding old billingAddress payment only listener and fixing it that way so this is no longer in the PR

**Review by njfaries — APPROVED** [review]

**Review by danac-gn — APPROVED** [review]

File: view/frontend/templates/js/replacejs.phtml [hunk 6]
```diff
@@ -406,6 +406,9 @@ if (!$block->isOnPageFromWhiteList() && !$block->isMinicartEnabled()) { return;
             });
         };
 
+        var isPaymentOnlyCheckout = function() {
```
Comment (vitaliyreznikov) [inline view/frontend/templates/js/replacejs.phtml:409]: I think if we refactor something it's better to change across the code.
After this change sometimes we use `isPaymentOnlyCheckout()` and sometimes `getCheckoutType()`.
I don't know how important it is but function getCheckoutType() return checkout for page like http://m2.vitaliy.dev.bolt.me/checkout/ and allow mini cart work on this page.

Can you explain a little why you need this change?

Comment (mikeshutov-bolt) [reply view/frontend/templates/js/replacejs.phtml:409]: It's a fair point. I'll explain my logic and why I went with this route. 

getCheckoutType - onload of /checkout page the url will not always have a # value when this function runs for native checkout. This results in an inconsistant result where most of the time it is actually 'checkout' and not 'payment' so conditions which depends on 'payment' being there onload will not work. The only situation when 'payment' was returned as value was when I refreshed with # values in the url. 

The only reason I did not modify the existing function is I am concerned about additional items breaking if I were to change it. However I am not against modifying getCheckoutType to a similar condition where we check for this:
```javascript
        var getCheckoutType = function() {
            var trimmedUrl = trim(location.pathname, '/');
            return trimmedUrl === 'checkout' && trimmedUrl !== 'checkout/cart' && settings.quote_is_virtual  ? 'payment' : 'checkout';
        };
```

Comment (vitaliyreznikov) [reply view/frontend/templates/js/replacejs.phtml:409]: We jumped on call and Mike explained me the fix. I think we can go ahead and update getCheckoutType function.
Can it be just
```
    var getCheckoutType = function() {
            return trim(location.pathname, '/') === 'checkout'   ? 'payment' : 'checkout';
        };
```

**Review by vitaliyreznikov — APPROVED** [review]
LGTM with one minor comment

File: view/frontend/templates/js/replacejs.phtml (hunk 6 above)
Comment (mikeshutov-bolt) [reply view/frontend/templates/js/replacejs.phtml:409]: Retesting with change.

**Comment by lbliii** [comment]
@matt-thomason, catching up on release notes and i was wondering if this bugfix is worth externalizing? (did it happen frequently enough to shoppers? it seems minor and i feel like we usually override most native checkout pages but just want to check if I understand the scenario fully).

**Pull request merged**